packages/runtime/src/ctx.ts

```diff
@@ -765,210 +787,246 @@ export class Ctx implements Context {
+							r = this.runtime.run(
+								appParams[APPROVAL_PROGRAM],
+								ExecutionMode.APPLICATION,
+								idx,
+								this.debugStack
+							);
+							this.deleteApp(signedTransaction.txn.appIndex);
+							break;
+						}
+						case algosdk.OnApplicationComplete.OptInOC: {
+							this.tx = this.gtxs[idx]; // update current tx to tx being exectuted in group
+							r = this.optInToApp(fromAccountAddr, signedTransaction.txn.appIndex, idx);
+							break;
+						}
+						case algosdk.OnApplicationComplete.UpdateApplicationOC: {
+							this.tx = this.gtxs[idx]; // update current tx to the requested index
+							if (appDefinition === undefined) {
+								throw new Error("Not supported");
+							}
+							r = this.updateApp(
+								signedTransaction.txn.appIndex,
+								appDefinition[idx] as SmartContract,
+								idx
+							);
+							break;
+						}
+					}
 					break;
 				}
-				case types.TransactionType.ClearApp: {
-					this.tx = this.gtxs[idx]; // update current tx to the requested index
-					const appParams = this.runtime.assertAppDefined(
-						txParam.appID,
-						this.getApp(txParam.appID)
-					);
-					try {
-						r = this.runtime.run(
-							appParams["clear-state-program"],
-							ExecutionMode.APPLICATION,
-							idx,
-							this.debugStack
+				case TransactionType.acfg: {
+					if (isEncTxAssetCreate(signedTransaction.txn.get_obj_for_encoding() as EncTx)) {
+						this.tx = this.gtxs[idx]; // update current tx to the requested index
+						const senderAcc = this.getAccount(fromAccountAddr);
+						const flags: ASADeploymentFlags = {
+							...payFlags,
+							creator: { ...senderAcc.account, name: senderAcc.address },
+						};
+						r = this.deployASADef(
+							signedTransaction.txn.assetName,
+							webTx.getTransactionASADefinition(signedTransaction.txn),
+							fromAccountAddr,
+							flags
 						);
-					} catch (error) {
-						// if transaction type is Clear Call,
-						// remove the app without throwing an error (rejecting tx)
-						// tested by running on algorand network
-						// https://developer.algorand.org/docs/features/asc1/stateful/#the-lifecycle-of-a-stateful-smart-contract
+						this.knowableID.set(idx, r.assetIndex);
+					} else if (
+						isEncTxAssetReconfigure(signedTransaction.txn.get_obj_for_encoding() as EncTx)
+					) {
+						const asset = this.getAssetDef(signedTransaction.txn.assetIndex);
+						if (asset.manager !== fromAccountAddr) {
+							throw new RuntimeError(RUNTIME_ERRORS.ASA.MANAGER_ERROR, {
+								address: asset.manager,
+							});
+						}
+						// modify asset in ctx.
+						r = this.modifyAsset(
+							signedTransaction.txn.assetIndex,
+							webTx.getAssetReconfigureFields(signedTransaction.txn)
+						);
+					} else if (
+						isEncTxAssetDeletion(signedTransaction.txn.get_obj_for_encoding() as EncTx)
+					) {
+						const asset = this.getAssetDef(signedTransaction.txn.assetIndex);
+						if (asset.manager !== fromAccountAddr) {
```

Comment:
I see we repeat this pattern many time. I think we should create verify function for this,